volatility3/framework/plugins/windows/suspended_threads.py

```diff
@@ -0,0 +1,148 @@
+            ),
+            requirements.VersionRequirement(
+                name="pe_symbols", component=pe_symbols.PESymbols, version=(1, 0, 0)
+            ),
+            requirements.VersionRequirement(
+                name="threads", component=threads.Threads, version=(1, 0, 0)
+            ),
+        ]
+
+    def _generator(self):
+        """
+        The goal of this plugin is to report on threads that are suspended
+
+        Legitimate programs can start threads suspended but then will later resume them
+
+        Subsets of malware techniques, such as EDR evasion and process hollowing,
+        create suspended threads and do not resume them. These are the threads that this
+        plugin is designed to catch.
+
+        See the whitepaper from our DEF CON 2024 presentation for more details:
+
+        https://www.volexity.com/wp-content/uploads/2024/08/Defcon24_EDR_Evasion_Detection_White-Paper_Andrew-Case.pdf
+        """
+        kernel = self.context.modules[self.config["kernel"]]
+
+        vads_cache: Dict[int, pe_symbols.PESymbols.ranges_type] = {}
+
+        proc_modules = None
+
+        # walk the threads of each process checking for suspended threads
+        for proc in pslist.PsList.list_processes(
+            context=self.context,
+            layer_name=kernel.layer_name,
+            symbol_table=kernel.symbol_table_name,
+        ):
+            for thread in threads.Threads.list_threads(kernel, proc):
+                try:
+                    # we only care if the thread is suspended
+                    if thread.Tcb.SuspendCount == 0:
+                        continue
+
+                    # 4 == terminated
+                    if thread.Tcb.State == 4:
+                        continue
+
+                    owner_proc = thread.owning_process()
+                    owner_proc_pid = thread.Cid.UniqueProcess
+                    owner_proc_name = utility.array_to_string(owner_proc.ImageFileName)
+                    thread_tid = thread.Cid.UniqueThread
+                    thread_start_addr = thread.StartAddress
+                    thread_win32_addr = thread.Win32StartAddress
+                except exceptions.InvalidAddressException:
+                    continue
+
+                # Nothing useful to report if a process doesn't have VADs.. Also a sign of smear/terminated
+                vads = pe_symbols.PESymbols.get_vads_for_process_cache(
+                    vads_cache, owner_proc
+                )
+                if not vads:
+                    continue
+
+                # Only compute this if needed as its expensive and 99.9% of samples
+                # will not have suspended threads
+                if not proc_modules:
+                    proc_modules = pe_symbols.PESymbols.get_process_modules(
+                        self.context, kernel.layer_name, kernel.symbol_table_name, None
+                    )
+
+                    path_and_symbol = partial(
```

Comment:
Would prefer this to be `functools.partial` and just `import partial` above.  The only exception we've made so far is for types, which are used *everywhere*.  I'd prefer not to let this open a gap for other functions to claim asylum please...

Reply:
Fixed the import method.

I want to leave the plugin as is for the generator as the real fix will be coming soon where `gather_thread_info` in thrdscan returns the information about the paths and symbols so that all the thread related plugins get it. Right now our threads output is pretty bare bone and not very usable in investigations. Once the update goes in then threads/thrdscan/<all threads plugins> will produce actionable output.